services/src/util/config.rs

```diff
@@ -0,0 +1,113 @@
+use std::sync::RwLock;
+
+use crate::error;
+use crate::error::Result;
+use config::{Config, File};
+use lazy_static::lazy_static;
+use serde::Deserialize;
+use snafu::ResultExt;
+use std::path::Path;
+
+lazy_static! {
+    static ref SETTINGS: RwLock<Config> = RwLock::new({
+        let mut settings = Config::default();
+        // test may run in subdirectory
+        #[cfg(test)]
+        let paths = [
+            "../Settings-default.toml",
+            "Settings-default.toml",
+            "../Settings-test.toml",
+            "Settings-test.toml",
+        ];
+
+        #[cfg(not(test))]
+        let paths = ["Settings-default.toml", "Settings.toml"];
+
+        #[allow(clippy::filter_map)]
+        let files: Vec<File<_>> = paths
+            .iter()
+            .map(Path::new)
+            .filter(|p| p.exists())
+            .map(File::from)
+            .collect();
+
+        settings.merge(files).unwrap();
+
+        settings
+    });
+}
+
+pub fn get_config<'a, T>(key: &str) -> Result<T>
+where
+    T: Deserialize<'a>,
+{
+    SETTINGS
+        .read()
+        .map_err(|_| error::Error::ConfigLockFailed)?
+        .get::<T>(key)
+        .context(error::Config)
+}
+
+pub fn get_config_element<'a, T>() -> Result<T>
+where
+    T: ConfigElement + Deserialize<'a>,
+{
+    SETTINGS
+        .read()
+        .map_err(|_| error::Error::ConfigLockFailed)?
+        .get::<T>(T::KEY)
+        .context(error::Config)
+}
+
+pub trait ConfigElement {
+    const KEY: &'static str;
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Web {
+    pub bind_address: String,
+    pub external_address: Option<String>,
+    pub backend: Backend,
+}
+
+impl ConfigElement for Web {
+    const KEY: &'static str = "web";
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Backend {
+    InMemory,
+    Postgres,
+}
+
+impl ConfigElement for Backend {
+    const KEY: &'static str = "backend";
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Postgres {
+    pub config_string: String,
+}
+
+impl ConfigElement for Postgres {
+    const KEY: &'static str = "postgres";
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProjectService {
+    pub list_limit: u32,
+}
+
+impl ConfigElement for ProjectService {
+    const KEY: &'static str = "project_service";
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GdalSource {
+    pub raster_data_root_path: String, // TODO: Path?
```

Comment:
yes paths everywhere

Reply:
I could use `PathBuf` as `Path` is not owned. However it doesn't seems to check if it's really a valid path in the file system. So makes no differences to using string

Reply:
i think PathBuf is fine since String is also an owned type.

Reply:
e44b19d69a1fa7d2452c8255a59f491b6b6353c3